### DISCUSSION TITLE
Using Contour Ingress Controller

### DISCUSSION THREAD

**Discussion by prateekkohli21** [strimzi/strimzi-kafka-operator#7733] [Q&A]
Hi,

Is it possible to use Contour Ingress Controller instead of [NGINX Ingress Controller for Kubernetes](https://github.com/kubernetes/ingress-nginx) for exposing kafka brokers.

**Comment by scholzj** [comment]
Strimzi is developed and tested with the Kubernetes Nginx Controller. Other Ingress controllers might work, might work with some manual steps or might not work. It is a bit hard to say since there is no good compatibility between them. Last time I tried Countour, it did not support neither TLS passthrough nor TCP routing - it supported only HTTP traffic. So it was useless for Strimzi. But that was a long time ago and I have no idea what the current state is - I expect Contour made a lot of progress since. So I'm afraid you will need to try it and see if it works or not.

What should work in any case assuming Contour now supports TLS passthrough or some form of TCP routing is the following:
* On Strimzi 0.32, you can create the `type: cluster-ip` listener in Kafka CR
* You can manually create the Ingress resources with whatever Contour needs and map them against the per-broker services
* And then in the listener configuration in the Kafka CR, you have to configure the `advertisedHost` and `advertisedPort` fields to the addresses of the Ingress resources so that the clients can find the brokers.

**Reply by prateekkohli21** [reply]
Will try this, thanks!!

**Reply by prateekkohli21** [reply]
Is there a way I can create per pod service in Strimzi 0.31.1 while using an internal listener (same as ClusterIP provided in Strimzi 0.32)?

**Reply by scholzj** [reply]
Well, you can always create an `internal` listener and then create the services manually. But there is no other way in Strimzi 0.31.

**Reply by prateekkohli21** [reply]
I tried creating the services manually with the same labels as the bootstrap service, but these services automatically get deleted.

I think the below code in KafkaListenersReconiciler.java is responsible for deleting these service, not sure though:
 // Delete services which match our selector but are not desired anymore
                    for (String serviceName : existingServiceNames) {
                        serviceFutures.add(serviceOperator.reconcile(reconciliation, reconciliation.namespace(), serviceName, null));
                    }

Is there any way I can stop these services from getting deleted?

**Reply by scholzj** [reply]
Because you used the same labels ;-). The labels tell Strimzi that these might be some left-over services that should have been deleted but were not. So it goes and deletes them. Without the labels, Strimzi should let them be.

**Reply by prateekkohli21** [reply]
Ohh right, it worked after removing the labels.

Thanks for that!!
